File: Sorting/quickSort.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <utility>

using namespace std;

int partition(vector<int> &arr, int low, int high)
{
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j <= high - 1; ++j)
    {
        if (arr[j] <= pivot)
        {
            ++i;
            swap(arr[i], arr[j]);
        }
    }

    swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quickSortIterative(vector<int> &arr, int low, int high)
{
    stack<pair<int, int>> stk;
    stk.push({low, high});

    while (!stk.empty())
    {
        pair<int, int> current = stk.top();
        stk.pop();
        low = current.first;
        high = current.second;

        if (low < high)
        {
            int pi = partition(arr, low, high);

            // Push elements before and after the pivot for sorting
            if (pi - 1 > low)
            {
                stk.push({low, pi - 1});
            }
            if (pi + 1 < high)
            {
                stk.push({pi + 1, high});
            }
        }
    }
}

int main()
{
    vector<int> arr = {64, 25, 12, 22, 11, 1};

    cout << "Original Array: ";
    for (int num : arr)
        cout << num << " ";
    cout << endl;

    int low = 0;
    int high = arr.size() - 1;
    quickSortIterative(arr, low, high);

    cout << "Sorted Array: ";
    for (int num : arr)
        cout << num << " ";
    cout << endl;

    return 0;
}

File: Sorting/quickSort.js
function swap(arr, i, j) {
    let temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

function partition(arr, p) {
    let { lb, ub } = p;
    let x = arr[lb];
    let down = lb + 1;
    let up = ub;

    while (down <= up) {
        while (down <= up && arr[down] <= x) {
            down++;
        }
        while (down <= up && arr[up] > x) {
            up--;
        }
        if (down < up) {
            swap(arr, down, up);
            down++;
            up--;
        }
    }
    swap(arr, lb, up);
    return up;
}

function quicksort(arr) {
    let stack = [];
    let p = { lb: 0, ub: arr.length - 1 };
    stack.push(p);

    while (stack.length > 0) {
        p = stack.pop();
        while (p.lb < p.ub) {
            let j = partition(arr, p);
            if (j - p.lb > p.ub - j) {
                stack.push({ lb: p.lb, ub: j - 1 });
                p.lb = j + 1;
            } else {
                stack.push({ lb: j + 1, ub: p.ub });
                p.ub = j - 1;
            }
        }
    }
}

let a = [20, 10, 3, 2, 1, 25, 35, 21, 23, 12, 11, -1, 44, 34, 23, 34, 56, -111, 34, 56];
quicksort(a);

console.log("sorted array:");
console.log(a.join(" "));
